### DISCUSSION TITLE
Welcome to trino-openapi Discussions!

### DISCUSSION THREAD

**Discussion by nineinchnick** [nineinchnick/trino-openapi#60] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by LocNguyend** [comment]
Is it required that the request parameter passed in must be a string, because when I pass a decimal number, I get this error: 'Unexpected constraint for confidentlevel_req(decimal)?'

**Reply by nineinchnick** [reply]
That depends on the type of that column, can you share the schema? Like the result of `SHOW CREATE TABLE foo`

**Reply by nineinchnick** [reply]
Maybe processing constraints of some types is not implemented yet.

**Reply by LocNguyend** [reply]
CREATE TABLE openapi_var.default.v3_var_ ( 
    confidentlevel decimal(18, 8) NOT NULL, 
    accountnumber varchar NOT NULL,         
    riskfreerate decimal(18, 8) NOT NULL,   
    var decimal(18, 8) NOT NULL,            
    typeofaccount varchar NOT NULL,         
    sharpe_ratio decimal(18, 8) NOT NULL,   
    lastdate date NOT NULL,                 
    accountnumber_req varchar,              
    confidentlevel_req decimal(18, 8),      
    riskfreerate_req decimal(18, 8),        
    ndays_var integer NOT NULL,             
    ndays_var_req integer,                  
    lastdate_req varchar,                   
    typeofaccount_req varchar               
 )        
 there are 6 fields required, i have tried to pass string to API then cast them into float and int, but its still not working

**Reply by nineinchnick** [reply]
Can you try a query like `SELECT * FROM openapi_var.default.v3_var_ WHERE confidentlevel = DECIMAL '1.0'`?

**Reply by nineinchnick** [reply]
Yeah we're missing support for decimals here: https://github.com/nineinchnick/trino-openapi/blob/main/src/main/java/pl/net/was/OpenApiClient.java#L232
It should be relatively easy to add this.

**Reply by nineinchnick** [reply]
I opened an issue to track this: https://github.com/nineinchnick/trino-openapi/issues/61

**Reply by LocNguyend** [reply]
My function takes three parameters, accountnumber, typeofaccount, and lastdate, which are passed as strings to query data from a table in the database. Then, these three parameters are passed as the remaining parameters (input by the user) to the API to perform calculations. Finally, the function returns a Pydantic model including the three remaining parameters that are not available in the database. So, do you have any suggestions on how to accomplish this? And if not decimal (float), what data type can be used in the Pydantic model?

**Reply by LocNguyend** [reply]
I have tried to convert the input from a string to perform calculations within the function and then cast it back to a string after the calculations. However, it seems that when passing it to a column with the suffix '_req', it is not possible to cast the string to a number for calculation. I encountered the following error: 
'class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')'

**Comment by LocNguyend** [comment]
> Yeah we're missing support for decimals here: https://github.com/nineinchnick/trino-openapi/blob/main/src/main/java/pl/net/was/OpenApiClient.java#L232 It should be relatively easy to add this.

Also can you put support for date datatype

**Reply by nineinchnick** [reply]
Done in efdf7354bf49ca5de0c1b8fccd3315daa9edb76a

**Comment by kandass12340** [comment]
Hello, 
I would like to use the Open API connector to retrieve (only select operations) data from a custom data structure that I have. How do I specify the table name and columns to this connector ?

**Reply by nineinchnick** [reply]
You provide it with an OpenAPI spec and the connector translates it into an SQL schema. There's no documentation yet on how it's done (#18) but you can look at the tests to get an idea what tables and columns will be created for a particular schema: https://github.com/nineinchnick/trino-openapi/blob/main/src/test/java/pl/net/was/TestOpenApiSpec.java#L111

Or just load up your spec and look around.

**Comment by vinodahub** [comment]
Hi @nineinchnick , 
Needed help. Im trying to deploy trino-openapi connector on starburst setup and the moment i assign catalog associated with this connector to cluster, its failing with below error.
2024-07-12T06:05:19.289Z        ERROR   main    io.trino.server.Server  pl/net/was/OpenApiPlugin has been compiled by a more recent version of the Java Runtime (class file version 66.0), this version of the Java Runtime only recognizes class file versions up to 65.0
java.lang.UnsupportedClassVersionError: pl/net/was/OpenApiPlugin has been compiled by a more recent version of the Java Runtime (class file version 66.0), this version of the Java Runtime only recognizes class file versions up to 65.0

Starburst version: 446.0.0

And as starburst require SPI complaint for connectors, how to make trino-openapi compatible with starburst. Please help as im not expert and tester.

**Reply by nineinchnick** [reply]
You have to build it manually with an older Java version and probably changing the Trino dependency

**Reply by vinodahub** [reply]
I'm getting below error when doing build using Java 21. Not sure how to overcome this.

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.5.0:enforce (default) on project trino-openapi:
[ERROR] Rule 5: org.codehaus.mojo.extraenforcer.dependencies.EnforceBytecodeVersion failed with message:
[ERROR] Found Banned Dependency: io.trino:trino-spi:jar:451
[ERROR] Use 'mvn dependency:tree' to locate the source of the banned dependencies.

**Reply by vinodahub** [reply]
I provided dep.trino.version to 446 and running "mvn clean package" did succeed.

**Reply by vinodahub** [reply]
Hi @nineinchnick ,
One more help, built from this repo using 446 trino version and java 21, using the zip binary to create connector in starburst cluster and creating catalog using openapi-generator server details(followed petstore from https://github.com/nineinchnick/trino-openapi/tree/main/src/test) and while trying to insert new pet action failing with error:

`Unsupported type decimal(18,8) (io.trino.spi.type.ShortDecimalType)`

**Reply by nineinchnick** [reply]
Decimal support was added in efdf7354bf49ca5de0c1b8fccd3315daa9edb76a, somewhere around 436. Usually I'd ask you to open an issue for this, but since you're using a custom build with a closed source deployment, I can't offer any support.
